### PR TITLE
Add Dataset.from_spark

### PR DESCRIPTION
**Pull request by maddiedawson**
Adds static method Dataset.from_spark to create datasets from Spark DataFrames.

This approach alleviates users of the need to materialize their dataframe---a common use case is that the user loads their dataset into a dataframe, uses Spark to apply some transformation to some of the columns, and then wants to train on the dataset.

https://github.com/huggingface/datasets/issues/5678

### CODE REVIEW

File: src/datasets/packaged_modules/spark/spark.py
Comment (lu-wang-dl) [inline src/datasets/packaged_modules/spark/spark.py:66]:
```suggestion
        raise ValueError("When using Dataset.from_spark, all workers should be able to access cache_dir")
```

File: src/datasets/arrow_dataset.py
Comment (maddiedawson) [reply src/datasets/arrow_dataset.py]: Done

File: src/datasets/packaged_modules/spark/spark.py
Comment (maddiedawson) [reply src/datasets/packaged_modules/spark/spark.py]: Done

File: src/datasets/arrow_dataset.py
Comment (maddiedawson) [reply src/datasets/arrow_dataset.py]: Done

**Comment by maddiedawson** [comment]
@mariosasko Would you or another HF datasets maintainer be able to review this, please?

**Review by albertvillanova — COMMENTED** [review]
Thanks for the work done, @maddiedawson.

Please note there was a previous PR adding this functionality:
- #5401

However the addition of this conversion was rejected by @lhoestq: see comment https://github.com/huggingface/datasets/pull/5401#issuecomment-1372277934
> Thinking more about it I don't really see how those two methods help in practice, since one can already do datasets <-> pandas <-> spark and those two methods don't add value over this.

**Comment by lhoestq** [comment]
Amazing ! Great job @maddiedawson 

Do you know if it's possible to also support writing to Parquet using the HF ParquetWriter if `file_format="parquet"` ?

Parquet is often used when people want to stream the data to train models - which is suitable for big datasets. On the other hand Arrow is generally used for local memory mapping with random access.

> Please note there was a previous PR adding this functionality

Am I right to say that it uses the spark workers to prepare the Arrow files ? If so this should make the data preparation fast and won't fill up the executor's memory as in the previously proposed PR

**Comment by maddiedawson** [comment]
Thanks for taking a look! Unlike the previous PR's approach, this implementation takes advantage of Spark mapping to distribute file writing over multiple tasks. (Also it doesn't load the entire dataset into memory :) )

Supporting Parquet here sgtm; I'll modify the PR.

I also updated the PR description with a common Spark-HF use case that we want to improve.

**Comment by maddiedawson** [comment]
Hey @albertvillanova @lhoestq , would one of you be able to re-review please? Thank you!

File: src/datasets/io/spark.py
Comment (lhoestq) [inline src/datasets/io/spark.py:22]: I guess you can modify `AbstractDatasetReader`'s typing to allow `None` here
```suggestion
            path_or_paths=None,
```

File: tests/test_arrow_dataset.py
Comment (lhoestq) [inline tests/test_arrow_dataset.py:3642]: Could you add a test to make sure that calling `from_spark` twice in a row with two different DataFrames results in two different Dataset objects ?

The `DatasetBuilder` uses a `cache_dir` with subdirectories defined after the name of the builder ("spark") and the arguments of the dataset to generate. Here since the `**kwargs` passed to `DatasetBuilder.__init__` won't change between two `from_spark`, it may reuse the same cache and return the same dataset.

If this is the case we can add a `df_id` parameter to the `SparkConfig`, which can be set to to be the python `id` of the Spark DataFrame for example (or any identifier of the Spark DataFrame).

**Review by lhoestq — COMMENTED** [review]
Good job with this, I left a few comment :)

File: src/datasets/arrow_dataset.py
Comment (maddiedawson) [reply src/datasets/arrow_dataset.py]: Done

File: src/datasets/arrow_dataset.py
Comment (maddiedawson) [reply src/datasets/arrow_dataset.py]: Done

File: src/datasets/io/spark.py
Comment (maddiedawson) [reply src/datasets/io/spark.py]: Done. My first follow-up task after this PR will be to add support for as_streaming_dataset

File: tests/test_arrow_dataset.py
Comment (maddiedawson) [reply tests/test_arrow_dataset.py]: Done

File: src/datasets/io/spark.py
Comment (maddiedawson) [reply src/datasets/io/spark.py]: Done

File: tests/test_arrow_dataset.py
Comment (maddiedawson) [reply tests/test_arrow_dataset.py:3643]: I modified the DatasetBuilder to cache based on the dataframe's semanticHash and added a test for it. This isn't perfect; e.g if a user generates a dataset from query "SELECT * FROM T", adds rows to the table, and then regenerates the dataset, the semantic hash for the dataframe will be the same, so the user will get the stale data. However, they can pass the force_download=True option to from_spark to force redownloading. Wdyt?

**Comment by maddiedawson** [comment]
@lhoestq this is ready for another pass! Thanks so much 🙏

**Comment by maddiedawson** [comment]
Friendly ping @lhoestq , also cc @polinaeterna who may be able to help take a look?

File: tests/test_arrow_dataset.py
Comment (lhoestq) [reply tests/test_arrow_dataset.py:3632]: Looks good ! (I was off the last few days, sorry for the late reply)

File: src/datasets/arrow_dataset.py
Comment (maddiedawson) [reply src/datasets/arrow_dataset.py]: Done

File: src/datasets/arrow_dataset.py
Comment (maddiedawson) [reply src/datasets/arrow_dataset.py]: Done

File: src/datasets/arrow_dataset.py
Comment (maddiedawson) [reply src/datasets/arrow_dataset.py]: Done

File: src/datasets/arrow_dataset.py
Comment (maddiedawson) [reply src/datasets/arrow_dataset.py:1251]: Done

File: tests/test_arrow_dataset.py
Comment (maddiedawson) [reply tests/test_arrow_dataset.py:3632]: Unfortunately no, the same df will generate a different a semantic hash in different Spark sessions.

File: src/datasets/packaged_modules/__init__.py
Comment (maddiedawson) [reply src/datasets/packaged_modules/__init__.py]: Thanks! Done

File: setup.py
Comment (maddiedawson) [reply setup.py]: Done

**Comment by lhoestq** [comment]
Merging `main` into this branch should fix the CI

**Review by lhoestq — APPROVED** [review]
Awesome ! I'd be interested in having someone else review it and then we're good to merge :)

**Comment by maddiedawson** [comment]
Thanks @lhoestq ! Is there a way for me to trigger the github workflow myself to triage the test failure? I'm not able to repro the test failures locally.

**Comment by maddiedawson** [comment]
There were two test issues in the workflow that I wasn't able to reproduce locally:

- Python 3.7: createDataFrame fails due to a pickling error. I modified the tests to instead write and read from json files
- Python 3.10: A worker crashes for unknown reasons. I modified the spark setup to explicitly specify local mode in case it was trying to do something else; let's see if that fixes the issue

**Comment by maddiedawson** [comment]
Also one more question @lhoestq when is the next datasets release? We're hoping this can make it in

**Comment by lhoestq** [comment]
I just re-ran the CI.
I think we can do a release right after this PR is merged ;)

File: tests/test_arrow_dataset.py
Comment (lu-wang-dl) [inline tests/test_arrow_dataset.py:3623]: Could we split the test to a different class so that we can put the spark creation into `setUp` function?

**Review by lu-wang-dl — APPROVED** [review]
LGTM. Left some minor comments.

File: tests/test_arrow_dataset.py
Comment (maddiedawson) [reply tests/test_arrow_dataset.py]: I reverted this change

**Comment by maddiedawson** [comment]
Thanks all! @lhoestq could we re-run CI again please? I think we have to disable this feature on python 3.7 due to the pickling error. The other failure was due to https://issues.apache.org/jira/browse/SPARK-30952 so I rewrote the df processing

File: src/datasets/io/spark.py
Comment (maddiedawson) [reply src/datasets/io/spark.py:27]: Done

File: src/datasets/filesystems/__init__.py
Comment (maddiedawson) [reply src/datasets/filesystems/__init__.py:58]: Done

**Comment by maddiedawson** [comment]
Thanks @lhoestq , this is ready for another CI run. I pinned the pyspark version to see if that fixes the pickling issue

**Comment by maddiedawson** [comment]
The remaining CI issues have been addressed! They were

- dill=0.3.1.1 is incompatible with cloudpickle, used by Spark. The min-dependency tests use this dill version, and those were failing. I added a skip-test annotation to skip Spark tests when using this dill version. This shouldn't be a production issue since if users are using that version of dill, they won't really be able to do anything with Spark anyway.
- One of the Spark APIs used in this feature (mapInArrow) is incompatible with Windows. I filed a Spark ticket for the team to investigate. For the tests, I added another annotation to skip Spark tests on Windows. In the next PR (adding streaming mode), we should be able to support Windows since that won't use mapInArrow.

I ran the CI on my forked branch: https://github.com/maddiedawson/datasets/pull/2 Everything passes except one instance of tests/test_metric_common.py::LocalMetricTest::test_load_metric_frugalscore; it looks like a flake.

@lhoestq granted that the CI passes here, is this ok to merge and release? We'd like to put out a blog post tomorrow to broadcast this to Spark users!

**Comment by maddiedawson** [comment]
Thanks @lhoestq ! Could you help take a look at the error please? Seems unrelated...

FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_map_multiprocessing_on_disk - NotADirectoryError: [WinError 267] The directory name is invalid: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmptfnrdj4x\\cache-5c5687cf5629c97a_00000_of_00002.arrow'
===== 1 failed, 2152 passed, 23 skipped, 20 warnings in 461.68s (0:07:41) =====

**Comment by yanzia12138** [comment]
python 3.9.2
Got an error  _pickle.PicklingError use Dataset.from_spark.

Did the dataset import load data from spark dataframe using multi-node Spark cluster
df = spark.read.parquet(args.input_data).repartition(50)
ds = Dataset.from_spark(df, keep_in_memory=True,
                        cache_dir="/pnc-data/data/nuplan/t5_spark/cache_data")
ds.save_to_disk(args.output_data)

Error : 
_pickle.PicklingError: Could not serialize object: RuntimeError: It appears that you are attempting to reference SparkContext from a broadcast variable, action, or transforma
tion. SparkContext can only be used on the driver, not in code that it run on workers. For more information, see SPARK-5063.
23/06/16 21:17:20 WARN ExecutorPodsWatchSnapshotSource: Kubernetes client has been closed (this is expected if the application is shutting down.)

**Comment by lhoestq** [comment]
Hi @yanzia12138 ! Could you open a new issue please and share the full stack trace ? This will help to know what happened exactly

**Pull request merged**
